### ISSUE TITLE
Order of checking for exit codes

### ISSUE THREAD

**Issue by greschd** [aiidateam/aiida-quantumespresso#413]
I am trying to add an exit code to ``PwCalculation`` for the case where the code crashes with a ``negative dexx`` error. The XML file doesn't exist in that case, and the standard output terminates with 

```
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
     Error in routine electrons (1):
     dexx is negative!  Check that exxdiv_treatment is appropriate for the system, or ecutfock may be too low
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

     stopping ...
```

I've added the exit code to the spec (in the calculation class), and in the ``message_map`` (https://github.com/aiidateam/aiida-quantumespresso/blob/develop/aiida_quantumespresso/parsers/parse_raw/pw.py#L34) added ``'dexx is negative': 'ERROR_DEXX_IS_NEGATIVE'`` as an error. In the parser itself, I've added a clause to ``validate_premature_exit`` (https://github.com/aiidateam/aiida-quantumespresso/blob/develop/aiida_quantumespresso/parsers/pw.py#L125).

The issue is that while the error shows up in the log
```
+-> ERROR at 2019-10-16 11:40:32.443091+00:00
 | parser returned exit code<300>: Both the stdout and XML output files could not be read or parsed.
+-> ERROR at 2019-10-16 11:40:32.439404+00:00
 | Both the stdout and XML output files could not be read or parsed.
+-> ERROR at 2019-10-16 11:40:32.433249+00:00
 | ERROR_OUTPUT_STDOUT_INCOMPLETE
+-> ERROR at 2019-10-16 11:40:32.428673+00:00
 | ERROR_DEXX_IS_NEGATIVE
+-> WARNING at 2019-10-16 11:40:32.423150+00:00
 | Error while parsing for energy terms.
```
it's still exit code ``300`` that is returned. I think that's because the check whether both XML and stdout had a failure in parsing happens before the ``validate_premature_exit`` call (https://github.com/aiidateam/aiida-quantumespresso/blob/develop/aiida_quantumespresso/parsers/pw.py#L100). 

Is this correct, or should the order be switched here? Maybe my approach to adding the exit code is wrong?

**Comment by sphuber** [comment]
No I think your approach is exactly as I intended it. Probably the check for errors in both stdout and XML should just be moved to afterwards or at least the logic should change. In this case we clearly have a more specific error then just "problems with both stdout and XML" so that should take precedence.

I think the problem here is that you might be using a version of QE using the new XML format which uses parsing based on a schema file. This is a lot more strict than the old XML parser and so the new one will fail whereas the old XML parser will simply have incomplete data. If you can add a test case to `tests/parsers/test_pw.py` where you check that you get the correct exit status, we can see how moving `validate_premature_exit` before the general check, changes affect the rest.

**Comment by greschd** [comment]
> I think the problem here is that you might be using a version of QE using the new XML format which uses parsing based on a schema file.

The QE version is 6.4.1, but since it doesn't produce any XML file I'm not sure how that changes things. Or does the parser detect which version was run from the stdout?

I'll see if I can come up with a good test case :+1:

**Comment by sphuber** [comment]
The test case would be simple, just create the folder `tests/parsers/fixtures/pw/failed_dexx_negative` and put the `aiida.out` file in there from your example run. Since you say the behavior of QE is to not even print an XML file, you don't have to add it. Then you just add a test function in `tests/parsers/test_pw.py` that just checks that the returned exit code is the one you expect

**Comment by sphuber** [comment]
Fixed in #439
